adal/src/main/java/com/microsoft/aad/adal/Oauth2.java

```diff
@@ -452,7 +470,26 @@ private AuthenticationResult postMessage(String requestMessage, Map<String, Stri
                 // Protocol related errors will read the error stream and report
                 // the error and error description
                 Logger.v(TAG, "Token request does not have exception");
-                result = processTokenResponse(response);
+                try {
+                    result = processTokenResponse(response);
+                } catch (final AuthenticationException e) {
+                    if (e.getCode().equals(ADALError.SERVER_ERROR_FOR_RETRY) && retryOnce) {
+                        //retry once if it is a server error
+                        //500, 503 and 504 are the ones we retry
+                        retryOnce = false;
+                        try {
+                            //retry once after half second
+                            final CountDownLatch signal = new CountDownLatch(1);
+                            signal.await(delayTimePeriod, TimeUnit.MILLISECONDS);
```

Comment:
I would like an answer to sleep vs latch as well, more for education. You just picked one is a fair answer as well.

---
In reply to: [70118292](https://github.com/AzureAD/azure-activedirectory-library-for-android/pull/676#discussion_r70118292) [](ancestors = 70118292)